cci_cameroon/analysis/text_similarity/similarity_search_sample.py

```diff
@@ -0,0 +1,608 @@
+
+# %%
+index2.add(sentence_embeddings)
+index2.ntotal  # number of embeddings indexed
+
+# %%
+# we now perform a search for similarity using the input text read earlier on.
+# %time
+distance2, position2 = index2.search(xq, k)  # search
+print(position2)
+print(distance2)
+
+# %%
+# get the text
+for i in position2[0]:
+    print(sentences[i])
+print(distance2[0])
+
+# %% [markdown]
+# ## By compressing the vectors, we further improve the computation time
+# Compression is implemented in the following cell by specifying the number of centroids and bits in each centroid
+
+# %%
+m = 8  # number of centroid IDs in final compressed vectors
+bits = 8  # number of bits in each centroid
+quantizer = faiss.IndexFlatL2(dimension)  # we keep the same L2 distance flat index
+index3 = faiss.IndexIVFPQ(quantizer, dimension, num_cells, m, bits)
+
+# %%
+index3.train(sentence_embeddings)
+index3.add(sentence_embeddings)
+
+# %%
+index3.nprobe = 1  # nprobe specifies the number of neigboring cells to be used
+
+# %%
+# we now perform a search for similarity using the input text read earlier on.
+# %time
+distance3, position3 = index3.search(xq, k)  # search
+# get the text
+for i in position3[0]:
+    print(sentences[i])
+print(distance3[0])
+
+# %%
+sentences[position[0][0]]
+
+# %%
+sentences[position[0][1]]
+
+# %%
+print("Similarity:", util.dot_score(xq, sentence_embeddings)[0][position[0][1]])
+
+# %% [markdown]
+# ## Output summary
+# For the given dataset, we notice that the algorithm returns the same set of indexes. However, when compression of the vectors happens, the indexes are returned in a reversed order. This shows that accuracy is compromised. The least distance between the vectors increases from 90 to 130.
+
+# %%
+# pip install hydra-core omegaconf
+
+# %% [markdown]
+# # Using the distilbert model for embedding
+
+# %%
+# using the distilbert model for embedding
+model2 = SentenceTransformer(
+    "distilbert-base-nli-stsb-mean-tokens"
+)  # best performance for all distillbert pretrained models.
+
+# %%
+# check GPU availability and set cuda if present.
+if torch.cuda.is_available():
+    model2 = model2.to(torch.device("cuda"))
+print(model2.device)
+
+# %%
+# produce sentence vectors
+embeddings = model2.encode(sentences)
+
+# %%
+# Step 1: Change data type
+embeddings2 = np.array([embedding for embedding in embeddings]).astype("float32")
+
+# %%
+print(embeddings[:10])
+
+# %%
+print(embeddings2[:10])
+
+# %%
+# create a faiss index for the search using the embenddings
+index_dis = faiss.IndexFlatL2(embeddings.shape[1])
+
+# %%
+index_dis.add(embeddings)  # add the vector to the index created
+
+# %%
+# search the index for our given input xq
+# %time
+distance4, position4 = index_dis.search(xq, k)  # search
+# get the text
+for i in position4[0]:
+    print(sentences[i])
+print(distance4[0])
+
+# %%
+sentences[position4[0][0]]
+
+# %%
+num_cells = 50  # Number of cells the index should be divided into
```

Comment:
Is there a particular reason for this number being 50? Should this number be stored in config (as its a hyper-parameter?).

https://nestauk.github.io/ds-cookiecutter/structure/#store-data-science-configuration-in-srcconfig